tools/swift-machine-learning-api/README.md

```diff
@@ -0,0 +1,336 @@
+            |__ utils
+                |__ __init__.py
+                |__ custom_utils_fn.py
+                |__ optimizer_utils.py
+                |__ metric_utils.py
+```
+
+## Instructions to run the code
+
+### Configuring the API
+
+- Create a config file named **_'config_file.yaml'_** like below, with your project id, bucket name and path of service account key along for main.py in the **config directory** for interacting with GCP resources.
+
+  ```
+  bucket_name: "gs://<bucket-name>"
+  project: <project_name>
+  project_id: <project_ID>
+  service_account_json_key: <path/of/service/account/key/json/file>
+  ```
+
+- Create a config file named **_'train.yaml'_** like below along for train.py in the **config directory** to specify train configurations. Refer to [*CMLE machine types*](https://cloud.google.com/ml-engine/docs/tensorflow/machine-types) for specifying type virtual machines you want the model to be trained. The trainer package will be created and updated in GCS bucket by running bash script update.sh. 
+
+    ```
+    scaleTier: "CUSTOM"
+    masterType: "complex_model_m"
+    workerType: "complex_model_m"
+    parameterServerType: "large_model"
+    workerCount: 3
+    parameterServerCount: 1
+    packageUris: ['<path/of/trainer/package/in/GCS>'] # bucket_name+'/trainer-0.0.tar.gz' 
+    region: "us-central1"
+    jobDir: <path/of/the/job_dir>
+    runtimeVersion: <tensorflow version>
+    pythonVersion: <python version> ##python2.7
+    ```
+- Create a config file named **_'config.yaml'_** like below in **codes** folder, with the type of virtual machines you want the model to be trained on. Currently if the training job is submitted through Train API the machine configurations are directly loaded from train.yaml. However to this has been added to provide an option to launch training jobs using [*command line*](https://cloud.google.com/sdk/gcloud/reference/ml-engine/jobs/submit/training)
+Example of  **_'config.yaml'_** looks like 
+  ```
+  trainingInput:
+    scaleTier: CUSTOM
+    masterType: complex_model_l
+    workerType: complex_model_l
+    parameterServerType: large_model
+    workerCount: 3
+    parameterServerCount: 1
+  ```
+
+- Check that your Storage and Machine Learning Engine API is enabled.
+- Create a service account key which should have following IAM roles:
+
+  > Machine Learning Engine > ML Engine Admin
+
+  > Storage > Storage Object Admin
+
+- Add the path of the Service account key to an environmental variable "GOOGLE_APPLICATION_CREDENTIALS"
+
+  ```bash
+  $ export GOOGLE_APPLICATION_CREDENTIALS=<path/of/service/account/key/json/file>
+  ```
+
+### Executing API
+
+- Upload/Update the trainer folder on the google cloud storage bucket using the following command. This command creates tar.gz file of the trainer directory and uploads it to the cloud storage bucket. Please ensure that "python" executable refers to python2.7. If no, please make the following change in update.sh 
```

Comment:
Is there a reason this needs to be python 2.7? We are trying to get most new assets to be python 3 due to the deprecation of 2.7 in 2020. However, if you explicitly depend on something (ie apache beam) which doesn't have python3 support we can make an exception.

Reply:
I double check with the team. The code actually can support python3. The only thing is that we should make sure the python version used to package is align with the python version selected on CMLE. I have modify the doc.